pandas/tests/frame/methods/test_first_and_last.py

```diff
@@ -10,38 +10,43 @@
 )
 import pandas._testing as tm
 
+deprecated_msg = "first is deprecated"
+
 
 class TestFirst:
     def test_first_subset(self, frame_or_series):
         ts = tm.makeTimeDataFrame(freq="12h")
         ts = tm.get_obj(ts, frame_or_series)
-        result = ts.first("10d")
-        assert len(result) == 20
+        with tm.assert_produces_warning(FutureWarning, match=deprecated_msg):
+            result = ts.first("10d")
 
-        ts = tm.makeTimeDataFrame(freq="D")
-        ts = tm.get_obj(ts, frame_or_series)
-        result = ts.first("10d")
-        assert len(result) == 10
+            assert len(result) == 20
 
-        result = ts.first("3M")
-        expected = ts[:"3/31/2000"]
-        tm.assert_equal(result, expected)
+            ts = tm.makeTimeDataFrame(freq="D")
+            ts = tm.get_obj(ts, frame_or_series)
+            result = ts.first("10d")
+            assert len(result) == 10
 
-        result = ts.first("21D")
-        expected = ts[:21]
-        tm.assert_equal(result, expected)
+            result = ts.first("3M")
+            expected = ts[:"3/31/2000"]
+            tm.assert_equal(result, expected)
 
-        result = ts[:0].first("3M")
-        tm.assert_equal(result, ts[:0])
+            result = ts.first("21D")
+            expected = ts[:21]
+            tm.assert_equal(result, expected)
+
+            result = ts[:0].first("3M")
+            tm.assert_equal(result, ts[:0])
 
     def test_first_last_raises(self, frame_or_series):
         # GH#20725
         obj = DataFrame([[1, 2, 3], [4, 5, 6]])
         obj = tm.get_obj(obj, frame_or_series)
 
         msg = "'first' only supports a DatetimeIndex index"
-        with pytest.raises(TypeError, match=msg):  # index is not a DatetimeIndex
-            obj.first("1D")
+        with tm.assert_produces_warning(FutureWarning, match=deprecated_msg):
+            with pytest.raises(TypeError, match=msg):  # index is not a DatetimeIndex
+                obj.first("1D")
```

Comment:
this is fine, though an alternative would be
```python
with (
    tm.assert_produces_warning(FutureWarning, match=deprecated_msg),
    pytest.raises(TypeError, match=msg)
):
    obj.first('1D')
```

There's a ruff check for this https://beta.ruff.rs/docs/rules/multiple-with-statements/ , in case you wanted to make a separate issue about it and we can enable it globally in a separate PR

Reply:
I opened a new issue.

Reply:
Imo this looks very ugly as soon as one of both statements gets longer. I would prefer keeping them separate for tests at least

Reply:
@phofl Separate like this you mean?:
```
with tm.assert_produces_warning(FutureWarning, match=deprecated_msg):
            with pytest.raises(TypeError, match=msg):  # index is not a DatetimeIndex
                obj.first("1D")
```